### PR TITLE
Refs #28459 -- Improved performance of select_related() when model is prefetched from its parent.

### PR DESCRIPTION
**Pull request by sir-sigurd**
https://code.djangoproject.com/ticket/28459

Before:
```
In [2]: %timeit for x in Human.objects.select_related('person'): pass
196 ms ± 2.05 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

After:
```
In [2]: %timeit for x in Human.objects.select_related('person'): pass
186 ms ± 1.59 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

### CODE REVIEW

File: django/db/models/query.py [hunk 1]
```diff
@@ -1683,10 +1683,7 @@ def __init__(self, klass_info, select, db):
             reorder_map.sort()
             self.init_list = [v[1] for v in reorder_map]
             pos_list = [row_pos for _, _, row_pos in reorder_map]
```
Comment (pope1ni) [inline django/db/models/query.py]: `pos_list` could be changed to be a generator now that it is only unpacked into `itemgetter`.

File: django/db/models/query.py [hunk 2]
```diff
@@ -1683,10 +1683,7 @@ def __init__(self, klass_info, select, db):
...
-
-            def reorder_for_init(row):
-                return [row[row_pos] for row_pos in pos_list]
-            self.reorder_for_init = reorder_for_init
+            self.reorder_for_init = operator.itemgetter(*pos_list)
```
Comment (pope1ni) [inline django/db/models/query.py]: `pos_list` could be changed to be a generator now that it is only unpacked into `itemgetter`:

```diff
-pos_list = [row_pos for _, _, row_pos in reorder_map]
-self.reorder_for_init = operator.itemgetter(*pos_list)
+self.reorder_for_init = operator.itemgetter(*(v[2] for v in reorder_map))
```

Comment (sir-sigurd) [inline django/db/models/query.py]: Probably it should be changed in another PR as it's not related to performance.

Comment (timgraham) [inline django/db/models/query.py]: It looks okay to me to do in the same commit.

Comment (ngnpope) [inline django/db/models/query.py]: I also noticed that in the first commit you did for this ticket there is ```operator.itemgetter(*[…])``` which could also be changed to use a generator to avoid the list instantiation.

Comment (sir-sigurd) [inline django/db/models/query.py]: It looks like it has worse performance:
```
In [9]: %timeit itemgetter(*[x for x in [1, 2, 3]])
435 ns ± 5.58 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
In [10]: %timeit itemgetter(*(x for x in [1, 2, 3])) 
694 ns ± 4.67 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
```
